tt_metal/tt_stl/any_range.hpp

```diff
@@ -0,0 +1,887 @@
+inline constexpr bool
+    range_compatible<TRange, TReference, TRangeTag, std::enable_if_t<std::is_base_of_v<sized_range_tag, TRangeTag>>> =
+        range_compatible<TRange, TReference, base_tag_t<TRangeTag>> and sized_range<TRange>;
+
+template <class T>
+using remove_cvref_t = std::remove_cv_t<std::remove_reference_t<T>>;
+
+template <class TIterator, class TReference, class TIteratorTag>
+inline constexpr bool iterator_compatible_cvref =
+    iterator_compatible<remove_cvref_t<TIterator>, TReference, TIteratorTag>;
+
+template <class TRangeTag, class TCanonicalTag, class TEnable = void>
+inline constexpr bool is_tag_v = std::is_base_of_v<TCanonicalTag, TRangeTag>;
+
+template <class TRangeTag, class TCanonicalTag>
+inline constexpr bool is_tag_v<TRangeTag, TCanonicalTag, std::void_t<base_tag_t<TRangeTag>>> =
+    std::is_base_of_v<TCanonicalTag, TRangeTag> and not std::is_base_of_v<TCanonicalTag, base_tag_t<TRangeTag>>;
+
+struct type_info {};
+
+template <class TValue>
+constexpr TValue *to_address(TValue *ptr) noexcept {
+    return ptr;
+}
+
+template <class TPointer>
+constexpr auto to_address(const TPointer &ptr) noexcept {
+    return detail::to_address(ptr.operator->());
+}
+
+class ICopyConstructible {
+   public:
+    [[nodiscard]] virtual const type_info *id() const = 0;
+    virtual void uninitialized_copy_to(void *dst_ptr) const = 0;
+    virtual void uninitialized_move_to(void *dst_ptr) noexcept = 0;
+    virtual ~ICopyConstructible() noexcept = default;
+};
+
+template <class TAnyIterator, class = iterator_category_t<TAnyIterator>>
+class IAnyIteratorAdaptor;
+
+template <class TAnyIterator>
+class IAnyIteratorAdaptor<TAnyIterator, std::input_iterator_tag> : public ICopyConstructible {
+   public:
+    [[nodiscard]] virtual bool operator==(const IAnyIteratorAdaptor &other) const = 0;
+    [[nodiscard]] virtual typename TAnyIterator::reference operator*() const = 0;
+    [[nodiscard]] virtual typename TAnyIterator::pointer operator->() const = 0;
+    virtual void operator++() = 0;
+};
+
+template <class TAnyIterator>
+class IAnyIteratorAdaptor<TAnyIterator, std::forward_iterator_tag>
+    : public IAnyIteratorAdaptor<TAnyIterator, std::input_iterator_tag> {
+   public:
+    using IAnyIteratorAdaptor<TAnyIterator, std::input_iterator_tag>::operator++;
+    [[nodiscard]] virtual TAnyIterator operator++(int) = 0;
+};
+
+template <class TAnyIterator>
+class IAnyIteratorAdaptor<TAnyIterator, std::bidirectional_iterator_tag>
+    : public IAnyIteratorAdaptor<TAnyIterator, std::forward_iterator_tag> {
+   public:
+    virtual void operator--() = 0;
+    [[nodiscard]] virtual TAnyIterator operator--(int) = 0;
+};
+
+template <class TAnyIterator>
+class IAnyIteratorAdaptor<TAnyIterator, std::random_access_iterator_tag>
+    : public IAnyIteratorAdaptor<TAnyIterator, std::bidirectional_iterator_tag> {
+   public:
+    virtual void operator+=(typename TAnyIterator::difference_type) = 0;
```

Comment:
It seems like according to https://en.cppreference.com/w/cpp/iterator/random_access_iterator, there should also be `operator-=(typename TAnyIterator::difference_type)` and `operator-(typename TAnyIterator::difference_type)`

Reply:
There is on the `stl::AnyIterator` and it delegates to this interface method. There's no reason to have a larger vtable than necessary.